driver-core/src/main/com/mongodb/internal/connection/AbstractMultiServerCluster.java

```diff
@@ -221,9 +223,9 @@ private void onChange(final ServerDescriptionChangedEvent event) {
                 oldClusterDescription = getCurrentDescription();
                 newClusterDescription = updateDescription();
             }
-        }
-        if (shouldUpdateDescription) {
-            fireChangeEvent(newClusterDescription, oldClusterDescription);
+            if (shouldUpdateDescription) {
```

Comment:
Move into synchronized block to ensure events are published in the order they are applied.